README
			-*- mode: auto-fill; coding: utf-8 -*-
taxbird - Gnome-2 based application to create Elster (Geier) XML files

Copyright (C) 2004,2005,2006 Stefan Siegl <stesie@brokenpipe.de>, Germany
                      (see AUTHORS for a complete and detailed list 
		       of authors, contributers and copyrightholders)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

The taxbird GUI as well as libgeier make use of the OpenSSL library, 
to which both the OpenSSL and the original SSLeay copyrights apply.
See README.OpenSSL for details. This product includes cryptographic
software written by Eric Young (eay@cryptsoft.com).

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Elster ist das deutsche Projekt zur elektronischen Abgabe
von Steuererklärungen, siehe <http://www.elster.de>.

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Taxbird ist ein Projekt, zur elektronischen Übermittlung von
Steuerdaten im Rahmen freier Software. Es besteht zum Einen aus
'libgeier', einer Bibliothek zur Übermittlung von Steuerdaten,
Sowie aus 'Taxbird' einer grafischen (auf Gnome-2 basierenden) 
Oberfläche zur Erfassung von Elster-XML Daten.

Mit Taxbird ist inzwischen die Übermittlung von Echtdaten möglich. Die
hierzu erforderliche Hersteller-ID wurde seitens der
Oberfinanzdirektion München (inzwischen umbenannt in `Bayerisches
Landesamt für Steuern) erteilt. Details zum Verfahren können in der
Datei README.vendor-id in diesem Verzeichnis nachgelesen werden.

Vorerst wird mittels Taxbird nur die Übermittlung von
Umsatzsteuervoranmeldungen möglich sein. Die Abgabe von
Lohnsteueranmeldungen ist vorgesehen und grundsätzlich machbar. Wer
das Formular benötigt und in Taxbird haben möchte, bitte auf der
Mailingliste melden. 

Die Übermittlung der Jahreserklärungen (hauptsächlich
Einkommensteuererklärungen) ist gegenwärtig noch nicht möglich, da die
Möglichkeit seitens der Finanzverwaltung bisher nicht eingeräumt
wurde. Voraussichtlich wird jedoch erst die Übermittlung der
Steuererklärungen für den Veranlagungszeitraum 2007 möglich sein. 

Die Formulare sind mittels Skriptsprache und Glade-XML in Taxbird
hinterlegt. So ist es möglich, ohne große Programmierkenntnisse eigene
(weitere) Formulare hinzuzufügen, zu installieren, sowie bestehende zu
Überarbeiten. Die Anwendung selbst muss nicht neu kompiliert werden.

Ab der Version 0.6 unterstützt Taxbird endlich auch die Abgabe von
digital signierten Steueranmeldungen. Näheres zum Verfahrensablauf
sowie, warum Sie diese Funktion unbedingt nutzen sollten, ist in der
Datei README.dsig erläutert.



Installation der CVS-Version:
-----------------------------

$ export CFLAGS="-Wall -ggdb" # sofern debugging Symbole gewünscht
$ autoreconf -i
$ ./configure
$ make
$ make install



Installation der Source-Tarball-Version:
----------------------------------------

$ ./configure
$ make
$ make install



Benötigte Bibliotheken:
-----------------------
GNU Guile 1.6.7 (andere können funktionieren)
OpenSSL 0.9.5 oder später
libgnomeui-2.0 (Gnome Development Libraries)
libgtkhtml-2.0 (GtkHTML Widget)



Zum Kompilieren der Anwendung werden benötigt:
----------------------------------------------
(andere können funktionieren oder auch nicht, bei Problemen bitte melden):

GNU Make 3.80
gcc (GCC) 3.3.5



Um den CVS-Tree kompilieren zu können, werden weiter benötigt:
--------------------------------------------------------------
(andere können funktionieren oder auch nicht, bei Problemen bitte melden):

autoconf (GNU Autoconf) 2.59
automake (GNU Automake) 1.4-p6
Glade-2 2.6.8 (mit Gnome-Unterstützung)



Teilnahmeerklärung
------------------

Bevor Sie mit Taxbird Echtdaten an das Finanzamt übermitteln können
(betroffen sind sowohl Lohnsteueranmeldungen als auch Umsatzsteuer-
voranmeldungen) ist eine sog. Teilnahmeerklärung abzugeben. Das hierzu
notwendige Formular können Sie auf den Seiten der Finanzverwaltung
unter https://www.elster.de/doc/teilnahmeerklaerung.pdf herunterladen.



(probably) Frequently Asked Questions
-------------------------------------
F: When running taxbird I sometimes get 
   "WARNING **: Invalid UTF8 string passed to pango_layout_set_text()".
   Why is this?

A: taxbird expects the Scheme sources to be encoded in UTF-8. No filtering
   etc. is done to the read strings, i.e. they are passed on to Gnome/Pango.
   If you encode your file with ISO-8859-1 for example, pango will not be able
   to interpret the data and will emit the warning. The shown strings will be
   truncated at the first non-ASCII character in this case.


F: Beim kompilieren tritt folgende Fehlermeldung auf:
   ** (glade-2:32360): WARNING **: Unknown widget class: GnomeApp

A: Sie haben zwar Glade-2 installiert, allerdings eine Variante, die
   keine Gnome-Unterstützung bietet. Unter Debian GNU/Linux ist zum
   Beispiel das Paket glade-gnome-2 zu installieren, nicht glade-2.


F: ./configure script complains:
   "Package openssl was not found in the pkg-config search path"

A: Sie haben entweder OpenSSL überhaupt nicht installiert, oder zumindest
   die zumindest die Datei openssl.pc Datei ist nicht im Pfad. Unter Debian
   GNU/Linux ist das Paket libssl-dev nebst Abhängigkeiten zu installieren.


F: In der Voransicht im Daten-Senden-Dialog sowie am ausgedruckten
   Protokoll sind die Felder "Unternehmer" sowie "Erstellt von" leer

A: Die Felder werden seitens dem von der Finanzverwaltung zur Verfügung
   gestellten Stylesheet nur ausgefüllt, wenn die Übermittlung durch 
   einen Berater erfolgt. Wenn das Feld Berufsbezeichnung
   (unter Datenlieferant) ausgefüllt wird, müssten die Daten erscheinen.


F: Nach einem Klick auf "Daten Senden" erscheint die Fehlermeldung
   "Die $Id: der Signatur ist nicht gültig (beim Prüfen der Signatur
    signatures/ustva-2005.sig)" - Export nur als Testdaten möglich

A: Der $Id: Wert der Signatur passt nicht zur installierten Programm-
   version von Taxbird. Vermutlich haben Sie noch eine "alte" GnuPG-
   Signatur installiert, die installierte Taxbird-Version (ab 0.2)
   erwartet jedoch eine OpenSSL-Signaturdatei.
   Sind unter $HOME/.taxbird/guile/ eventuell noch alte Skriptdaten
   hinterlegt? Falls ja, bitte löschen Sie diese.


Vielen Dank, dass Sie Taxbird verwenden!

$Id: README,v 1.12 2006-05-18 18:33:07 stesie Exp $


debian/changelog
taxbird (0.7-1) unstable; urgency=low

  * new upstream release
 
 -- Stefan Siegl <stesie@brokenpipe.de>  Sat, 13 May 2006 17:19:23 +0200

taxbird (0.6-5) unstable; urgency=low

  * register application/x-taxbird mimetime
  * automatically add .txb suffix
  * added .desktop file to automatically add entries to gui menus

 -- Stefan Siegl <stesie@brokenpipe.de>  Wed, 11 Jan 2006 22:02:23 +0100

taxbird (0.6-4) unstable; urgency=low

  * repackaging with a Debian/stable distribution
 
 -- Stefan Siegl <stesie@brokenpipe.de>  Tue, 06 Dec 2005 23:16:23 +0100

taxbird (0.6-3) unstable; urgency=low

  * new upstream release 0.6 (alias Santa Clause)

 -- Stefan Siegl <stesie@brokenpipe.de>  Tue, 06 Dec 2005 19:23:23 +0100

taxbird (0.6-2) unstable; urgency=low

  * pulled in ustva-2006 and ustsvza-2006 forms from cvs
  * added README.dsig

 -- Stefan Siegl <stesie@brokenpipe.de>  Sat, 26 Nov 2005 14:24:24 +0100

taxbird (0.6-1) unstable; urgency=low

  * cvs snapshot, featuring taxbird with digital signatures

 -- Stefan Siegl <stesie@brokenpipe.de>  Thu, 24 Nov 2005 21:22:27 +0100

taxbird (0.5-2) unstable; urgency=low

  * upstream release 0.5

 -- Stefan Siegl <stesie@brokenpipe.de>  Sat, 19 Nov 2005 19:20:38 +0100

taxbird (0.5-1) unstable; urgency=low

  * new upstream pre-release, this is just an cvs snapshot
  * new sheet definition format (based on Glade XML)
  * person-related data can be stored now
  * print post-processing fixed

 -- Stefan Siegl <stesie@brokenpipe.de>  Sat, 19 Nov 2005 12:38:39 +0100

taxbird (0.4-1) unstable; urgency=low

  * new upstream release, various minor bug fixes

 -- Stefan Siegl <stesie@brokenpipe.de>  Sat, 8 Oct 2005 20:50:24 +0200

taxbird (0.3-1) unstable; urgency=low

  * display success message when data has been sent successfully
  * directly link against libgeier instead of forking off command line tool
  * check for returned error messages

 -- Stefan Siegl <stesie@brokenpipe.de>  Sat, 23 Jul 2005 13:50:49 +0200

taxbird (0.2-1) unstable; urgency=low

  * Initial release

 -- Stefan Siegl <stesie@brokenpipe.de>  Tue, 21 Jun 2005 13:08:39 +0200



src/export.c
/* Copyright(C) 2005,2006 Stefan Siegl <stesie@brokenpipe.de>
 * taxbird - free program to interface with German IRO's Elster/Coala
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <gnome.h>
#include <libgtkhtml/gtkhtml.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <unistd.h>

#include <geier.h>

#include "interface.h"
#include "dialog.h"
#include "support.h"
#include "export.h"
#include "guile.h"
#include "form.h"

/* ask current template's guile backend to generate the data we'd
 * like to write out (in case testcase is set, with data being marked
 * as testcase data) */
static SCM taxbird_export_call_backend(GtkWidget *appwin, int testcase);

/* fork a subprocess, RETURN: -1 on error, 0 == child, pid == parent */
static pid_t taxbird_export_launch_subproc(int *to, int *from);

typedef pid_t (* taxbird_export_subproc) (int *to, int *from);

/* ask the user whether the exported data is okay */
static void taxbird_export_ask_user(GtkWidget *appwin, HtmlDocument *doc,
				    SCM data, SCM proto_fn);

/* launch the given command CMD (whether CMD may even contain arguments)
 * RETURN: pid of child on success, -1 on failure */
static pid_t taxbird_export_launch_command(const char *cmd, int *to, int *fr);


static int
taxbird_xsltify_text(geier_context *context,
		     const unsigned char *input, size_t inlen,
		     unsigned char **output, size_t *outlen)
{
  int retval;
  xmlDoc *indoc;
  xmlDoc *outdoc;

  if((retval = geier_text_to_xml(context, input, inlen, &indoc)))
    goto out0;

  if((retval = geier_xsltify(context, indoc, &outdoc)))
    goto out1;
  
  if((retval = geier_xml_to_encoded_text(context, outdoc, "ISO-8859-1",
					 output, outlen)))
    goto out2;

 out2:
  xmlFreeDoc(outdoc);
 out1:
  xmlFreeDoc(indoc);
 out0:
  return retval;
}


void
taxbird_export(GtkWidget *appwin, int testcase)
{
  SCM data = taxbird_export_call_backend(appwin, testcase);

  if(SCM_FALSEP(scm_list_p(data)))
    return; /* exporter function didn't return a list, thus abort here.
	     * error messages should have been emitted by the called func */

  /* export data into a c-string *********************************************/
  taxbird_guile_eval_file("xml-writer.scm");
  data = scm_call_1(scm_c_lookup_ref("xml-writer:export"), data);

  g_return_if_fail(SCM_STRINGP(data));
  const char *data_text = SCM_STRING_CHARS(data);
  const int data_text_len = strlen(data_text);

  /* initialize libgeier */
  geier_context *ctx = geier_context_new();
  if(! ctx) {
    taxbird_dialog_error(appwin, _("Unable to initialize libgeier context"));
    return;
  }

  /* try to validate resulting xml document against schema file */
  if(geier_validate_text(ctx, geier_format_unencrypted,
			 data_text, data_text_len)) {
    taxbird_dialog_error(appwin, _("Unable to validate the "
				   "exported document. Sorry, but this "
				   "should not happen. \n\nPlease consider "
				   "contacting the taxbird@taxbird.de "
				   "mailing list."));
    geier_context_free(ctx);
    return;
  }
	
  /* xsltify gathered data */
  unsigned char *data_xslt;
  size_t data_xslt_len;
  if(taxbird_xsltify_text(ctx, data_text, data_text_len,
			  &data_xslt, &data_xslt_len)) {
    taxbird_dialog_error(appwin, _("Unable to xsltify the "
				   "exported document. Sorry, but this "
				   "should not happen. \n\nPlease consider "
				   "posting to the taxbird@taxbird.de "
				   "mailing list."));
    geier_context_free(ctx);
    return;
  }

  geier_context_free(ctx);

  /* now convert xslt-result to HtmlDocument */
  HtmlDocument *doc = html_document_new();
  if(! doc) {
    taxbird_dialog_error(appwin, _("Unable to allocate HtmlDocument."));
    return;
  }

  if(! html_document_open_stream(doc, "text/html")) {
    html_document_clear(doc);
    taxbird_dialog_error(appwin, _("Unable to call open_stream"
				   "on allocated HtmlDocument."));
    return;
  }

  html_document_write_stream(doc, data_xslt, data_xslt_len);
  html_document_close_stream(doc);

  free(data_xslt);

  /* automatically fill protocol-store widget (which isn't filled by glade) **/
  taxbird_guile_eval_file("taxbird.scm");
  SCM databuf = (SCM) g_object_get_data(G_OBJECT(appwin), "scm_data");
  SCM fn = scm_call_1(scm_c_lookup_ref("tb:get-proto-filename"), databuf);

  /* ask the user, whether it's okay to send the data ************************/
  taxbird_export_ask_user(appwin, doc, data, fn);
  return;
}



/* export bottom half,
 * return 0 on success (i.e. destroy druid widget), 1 on error, a suitable
 * error message is emitted already */
int
taxbird_export_bottom_half(GtkWidget *confirm_dlg)
{
  SCM data = (SCM)g_object_get_data(G_OBJECT(confirm_dlg), "data");
  g_return_val_if_fail(SCM_STRINGP(data), 1);

  const char *data_text = SCM_STRING_CHARS(data);
  int data_text_len = strlen(data_text);

  geier_context *ctx = geier_context_new();
  if(! ctx) {
    taxbird_dialog_error(confirm_dlg,
			 _("Unable to initialize libgeier context."));
    return 1;
  }

  /* digitally sign the Coala-XML if requested *******************************/
  GtkWidget *dsig = lookup_widget(confirm_dlg, "dsig");
  if(gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(dsig))) {
    GtkEntry *e;

    e = GTK_ENTRY(lookup_widget(confirm_dlg, "dsig_fileentry_text"));
    const char *filename = gtk_entry_get_text(e);

    e = GTK_ENTRY(lookup_widget(confirm_dlg, "dsig_pincode"));
    const char *pincode = gtk_entry_get_text(e);
    
    unsigned char *output;
    size_t outlen;
    if(geier_dsig_sign_text(ctx, data_text, data_text_len, &output, &outlen,
			    filename, pincode)) {
      taxbird_dialog_error(confirm_dlg, 
			   _("Unable to digitally sign Elster document."));
      geier_context_free(ctx);
      return 1;
    }

    data = scm_take_str(output, outlen);
    data_text = SCM_STRING_CHARS(data);
    data_text_len = outlen;
  }       

  /* dump the generated Coala-XML to a file, if requested ********************/
  GtkWidget *store = lookup_widget(confirm_dlg, "store");
  if(gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(store))) {
    GtkEntry *e = GTK_ENTRY(lookup_widget(confirm_dlg, "store_entry"));
    const char *filename = gtk_entry_get_text(e);

    if(! filename) {
      taxbird_dialog_error(confirm_dlg, _("You told to store Coala-XML to "
					  "a file, however didn't specify "
					  "a filename."));
      geier_context_free(ctx);
      return 1;
    }

    int fd = open(filename, O_WRONLY | O_CREAT | O_TRUNC,
		  S_IRUSR | S_IRGRP | S_IROTH | S_IWUSR | S_IWGRP | S_IWOTH);
    if(fd < 0) {
      taxbird_dialog_error(confirm_dlg, _("Unable to create Coala-XML file."));
      geier_context_free(ctx);
      return 1;
    }

    write(fd, data_text, data_text_len);
    close(fd);
  }



  /*** check whether we are requested to send the data to the IRO ************/
  GtkWidget *send = lookup_widget(confirm_dlg, "send");
  if(! gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(send))) {
    geier_context_free(ctx);
    return 0; /* close the druid, we shalt not send the data */
  }

  /*** create filedescriptor for protocol output *****************************/
  pid_t pid_print = 0;
  int fd_to_lpr = -1, fd_from_lpr = -1;
  GtkWidget *print = lookup_widget(confirm_dlg, "protocol_print");
  if(gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(print))) {
    /* fork print process ***************/
    GtkWidget *w = lookup_widget(confirm_dlg, "protocol_print_command");
    const char *cmd = gtk_entry_get_text(GTK_ENTRY(w));

    pid_print = taxbird_export_launch_command(cmd, &fd_to_lpr, &fd_from_lpr);
    if(pid_print < 1) {
      taxbird_dialog_error(confirm_dlg, _("Unable to fork print process, "
					  "please verify the specified "
					  "command"));
      geier_context_free(ctx);
      return 1;
    }    
  }

  int fd_to_protofile = -1;
  GtkWidget *protosave = lookup_widget(confirm_dlg, "protocol_store");
  if(gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(protosave))) {
    /* open file to write protocol to ***/
    GtkWidget *w = lookup_widget(confirm_dlg, "protocol_store_fileentry_text");
    const char *filename = gtk_entry_get_text(GTK_ENTRY(w));

    fd_to_protofile = open(filename, O_WRONLY | O_CREAT | O_TRUNC,
			   S_IRUSR | S_IRGRP | S_IROTH | 
			   S_IWUSR | S_IWGRP | S_IWOTH);

    if(fd_to_protofile < 0) {
      taxbird_dialog_error(confirm_dlg, _("Unable to create protocol file."));
      geier_context_free(ctx);
      return 1;
    }
  }

  /* send the data, finally **************************************************/
  unsigned char *data_return;
  size_t data_return_len;
  if(geier_send_text(ctx, data_text, data_text_len,
		     &data_return, &data_return_len)) {
    taxbird_dialog_error(confirm_dlg,
			 _("Unable to send the exported document."));

    if(fd_from_lpr >= 0) close(fd_from_lpr);
    if(fd_to_lpr >= 0) close(fd_to_lpr);
    if(fd_to_protofile >= 0) close(fd_to_protofile);
    return 1;
  }

  char *error_msg = geier_get_clearing_error_text(ctx, data_return,
						  data_return_len);
  if(error_msg) {
    char *msg = g_strdup_printf(_("Unable to send the tax declaration to the "
				  "inland revenue office, the error message, "
				  "returned by the clearing host was: \n\n%s"),
				error_msg);
    taxbird_dialog_error(confirm_dlg, msg);
    free(msg);
    free(data_return);
    if(fd_from_lpr >= 0) close(fd_from_lpr);
    if(fd_to_lpr >= 0) close(fd_to_lpr);
    if(fd_to_protofile >= 0) close(fd_to_protofile);
    return 1;
  }

  /** now run xsltify to generate protocol for user */
  unsigned char *data_xslt;
  size_t data_xslt_len;
  if(taxbird_xsltify_text(ctx, data_return, data_return_len,
			  &data_xslt, &data_xslt_len)) {
    taxbird_dialog_error(confirm_dlg, _("Unable to generate the transmission "
					"protocol. This is a local error, THIS "
					"IS, YOUR DATA HAS BEEN SENT TO THE "
					"INLAND REVENUE OFFICE AND IT WAS "
					"ACCEPTED."));
    free(data_return);
    if(fd_from_lpr >= 0) close(fd_from_lpr);
    if(fd_to_protofile >= 0) close(fd_to_protofile);
    if(fd_to_lpr >= 0) close(fd_to_lpr);
    return 1;
  }

  if(fd_to_lpr >= 0) {
    write(fd_to_lpr, data_xslt, data_xslt_len);
    close(fd_to_lpr);
  }

  if(fd_to_protofile >= 0) {
    write(fd_to_protofile, data_xslt, data_xslt_len);
    close(fd_to_protofile);
  }

  /*** wait for lpr process to exit ******************************************/
  if(fd_from_lpr >= 0) {
    int exit_status;
    unsigned char buf[256];
    while(read(fd_from_lpr, buf, sizeof(buf)));

    close(fd_from_lpr);

    if((waitpid(pid_print, &exit_status, 0) != pid_print)
       || (! WIFEXITED(exit_status))
       || (WEXITSTATUS(exit_status))) {
      taxbird_dialog_error(confirm_dlg,
			   _("Printer process exited abnormally, "
			     "your data has been sent however."));
      return 1;
    }
  }

  taxbird_dialog_info(NULL, _("Your data has been sent to the inland "
			      "revenue office and was accepted"));
  return 0; /* hey, we did it. :) */
}


/* ask current template's guile backend to generate the data we'd
 * like to write out */
static SCM
taxbird_export_call_backend(GtkWidget *appwin, int test) 
{
  int current_form = (int) g_object_get_data(G_OBJECT(appwin), "current_form");
  g_return_val_if_fail(current_form >= 0, SCM_BOOL(0));

  /* data as supplied by the user */
  SCM data = (SCM) g_object_get_data(G_OBJECT(appwin), "scm_data");

  /* call exporter function of current template */
  return scm_call_2(forms[current_form]->dataset_export, data, SCM_BOOL(test));
}



/* launch the given command CMD (whether CMD may even contain arguments)
 * RETURN: pid of child on success, -1 on failure */
static pid_t 
taxbird_export_launch_command(const char *cmd, int *to, int *from)
{
  char **array = NULL;
  const char *start = cmd, *p = cmd;
  int quote = 0, len = 0;
  for(;; p ++) {
    if(*p == '\\') {
      g_return_val_if_fail(p[1] != '0', -1);
      p ++;
    }
    else if(*p == '\'' || *p == '\"') {
      if(quote && *p == quote) quote = 0; /* reset */
      else if(! quote) quote = *p; /* set */
    }
    else if(*p == ' ' || *p == '\t' || *p == '\0') {
      array = g_realloc(array, sizeof(char *) * (len + 1));
      array[len ++] = g_strndup(start, p - start);
      start = p + 1;
      if(! *p) break;
    }
  }
   
  array = g_realloc(array, sizeof(char *) * (len + 1));
  array[len] = NULL; /* terminate array */

  pid_t child = taxbird_export_launch_subproc(to, from);
  if(child != 0) {
    do
      g_free(array[-- len]);
    while(len);

    g_free(array);
    return child; /* main process || error -> get out */
  }

  /** launch subprocess */
  g_printerr("executing %s ...\n", array[0]);
  execvp(array[0], array);

  g_printerr("failed.\n");
  /** if execlp fails, die. */
  exit(255);
}



/* fork a subprocess, RETURN: -1 on error, 0 == child, pid == parent */
static pid_t 
taxbird_export_launch_subproc(int *to, int *from)
{
  int to_pipe[2];
  int from_pipe[2];

  if(pipe(to_pipe)) {
    perror(PACKAGE_NAME);
    return -1;
  }

  if(pipe(from_pipe)) {
    perror(PACKAGE_NAME);

    close(to_pipe[0]); /* close the pipe we've already set up */
    close(to_pipe[1]);

    return -1; /* good bye. */
  }

  pid_t child = fork();

  if(child < 0) {
    perror(PACKAGE_NAME);
    return -1;
  }

  if(child) {
    close(to_pipe[0]);   /* we don't read from the to-pipe */
    close(from_pipe[1]); /* we don't write to the from-pipe */

    *to = to_pipe[1];
    *from = from_pipe[0];

    return child; /* main process, return the pid */
  }

  /** this is the child process ... */
  close(to_pipe[1]);
  close(from_pipe[0]);

  /** map the pipe to stdin/stdout */
  if(dup2(to_pipe[0], 0) == -1) exit(255);
  if(dup2(from_pipe[1], 1) == -1) exit(255);

  return 0;
}



/* show the export druid and return */
static void
taxbird_export_ask_user(GtkWidget *appwin, HtmlDocument *doc, SCM data, SCM fn)
{
  (void) appwin;  /* unused for the moment */

  GtkWidget *confirm_dlg = create_dlgExportConfirmation();

  GtkWidget *htmlview = lookup_widget(confirm_dlg, "htmlview");
  html_view_set_document(HTML_VIEW(htmlview), doc);

  g_object_set_data_full(G_OBJECT(confirm_dlg), "data", 
			 (void*) scm_gc_protect_object(data),
			 (GDestroyNotify) scm_gc_unprotect_object);

  if(SCM_STRINGP(fn)) {
    GtkWidget *w = lookup_widget(confirm_dlg, "protocol_store_fileentry_text");
    gtk_entry_set_text(GTK_ENTRY(w), SCM_STRING_CHARS(fn));
  }

  gtk_widget_show(confirm_dlg);
}



